Sources/SecureXPC/Client/XPCClient.swift

```diff
@@ -157,87 +157,153 @@ public class XPCClient {
+        var connection: xpc_connection_t?
+        do {
+            connection = try getConnection()
+        } catch {
+            if let error = error as? XPCError {
+                handler(.failure(error))
+            } else {
+                handler(.failure(.unknown))
+            }
+        }
+        guard let connection = connection else {
+            return
+        }
+        
+        // Async send the message over XPC
+        xpc_connection_send_message_with_reply(connection, encoded, nil) { reply in
             let result: Result<R, XPCError>
-            if xpc_get_type(xpcResponse) == XPC_TYPE_DICTIONARY {
+            if xpc_get_type(reply) == XPC_TYPE_DICTIONARY {
                 do {
-                    let response = try Response(dictionary: xpcResponse)
+                    let response = try Response(dictionary: reply)
                     if response.containsPayload {
-                        result = Result.success(try response.decodePayload(asType: R.self))
+                        result = .success(try response.decodePayload(asType: R.self))
                     } else if response.containsError {
-                        result = Result.failure(try response.decodeError())
+                        result = .failure(try response.decodeError())
+                    } else if R.self == EmptyResponse.self { // Special case for when an empty response is expected
+                        result = .success(EmptyResponse.instance as! R)
                     } else {
-                        result = Result.failure(.unknown)
+                        result = .failure(.unknown)
                     }
                 } catch let error as XPCError  {
-                    result = Result.failure(error)
+                    result = .failure(error)
                 } catch {
-                    result = Result.failure(.unknown)
+                    result = .failure(.unknown)
                 }
-            } else if xpc_equal(xpcResponse, XPC_ERROR_CONNECTION_INVALID) {
-                result = Result.failure(.connectionInvalid)
-            } else if xpc_equal(xpcResponse, XPC_ERROR_CONNECTION_INTERRUPTED) {
-                result = Result.failure(.connectionInterrupted)
+            } else if xpc_equal(reply, XPC_ERROR_CONNECTION_INVALID) {
+                result = .failure(.connectionInvalid)
+            } else if xpc_equal(reply, XPC_ERROR_CONNECTION_INTERRUPTED) {
+                result = .failure(.connectionInterrupted)
             } else { // Unexpected
-                result = Result.failure(.unknown)
+                result = .failure(.unknown)
             }
-            self.handleConnectionErrors(event: xpcResponse)
-            reply(result)
-        })
+            self.handleConnectionErrors(event: reply)
+            handler(result)
+        }
     }
-
+    
+    /// Wrapper that handles responses without a payload since `Void` is not `Decodable`
```

Comment:
This is a real fly in the ointment, eh?

I got around this in my project by making a separate API whose error handler is `(XPCError?) -> Void`. I think the usability is a *tiny* bit nicer at the call-site, but it does make it less consistent with the other APIs.

Reply:
Oh hmm, so I originally wrote it up exactly like that but decided against it as it felt less clear because my intention wasn't to be an _error_ handler, but a _completion_ handler. So with this design pattern you can check for success instead of needing to check for the non-presence of an error. They are equivalent, but it just seemed a bit clearer to do it this way.